File: server/lib/export/templates/csharp/TheseusClient.cs
// Theseus-generated inference client.
//
// Wraps the exported ONNX model with the preprocessing/postprocessing Ludwig
// applied at training time, read from preprocessing.json — Ludwig's exported
// ONNX graph is the bare model only (no resize/normalize, no class-index
// decoding), so this client reconstructs that step from the manifest rather
// than assuming a fixed tensor layout: input/output tensor names are matched
// against the graph's actual names at runtime, not hardcoded, since ONNX
// export doesn't guarantee a stable naming convention across Ludwig versions.
//
// Currently implements preprocessing for `image` inputs with `category`
// outputs (image classification) — the modality this export path has been
// verified against. Other input types throw with a pointer to
// preprocessing.json so you can adapt preprocessing yourself; see README.md.

using System.Text.Json;
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;

namespace TheseusExport;

internal sealed class ImageNormalization
{
    public float[] Mean { get; set; } = [];
    public float[] Std { get; set; } = [];
}

internal sealed class PreprocessingInput
{
    public string Type { get; set; } = "";
    public string Column { get; set; } = "";
    public Dictionary<string, JsonElement>? LudwigPreprocessing { get; set; }
    public ImageNormalization? ImageNormalization { get; set; }
}

internal sealed class PreprocessingOutput
{
    public string Column { get; set; } = "";
    public List<string>? Classes { get; set; }
}

internal sealed class PreprocessingManifest
{
    public List<PreprocessingInput> Inputs { get; set; } = [];
    public List<PreprocessingOutput> Outputs { get; set; } = [];
}

public sealed class TheseusClient : IDisposable
{
    private static readonly JsonSerializerOptions JsonOpts = new() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };

    private readonly InferenceSession _session;
    private readonly PreprocessingInput _inputSpec;
    private readonly PreprocessingOutput _outputSpec;
    private readonly string _onnxInputName;
    private readonly string _onnxOutputName;

    public TheseusClient(string? modelPath = null, string? preprocessingPath = null)
    {
        var here = AppContext.BaseDirectory;
        modelPath ??= Path.Combine(here, "model.onnx");
        preprocessingPath ??= Path.Combine(here, "preprocessing.json");

        var manifest = JsonSerializer.Deserialize<PreprocessingManifest>(File.ReadAllText(preprocessingPath), JsonOpts)
            ?? throw new InvalidDataException("preprocessing.json is empty or invalid");
        _inputSpec = manifest.Inputs[0];
        _outputSpec = manifest.Outputs[0];

        _session = new InferenceSession(modelPath);
        _onnxInputName = MatchTensorName(_session.InputMetadata.Keys, _inputSpec.Column);
        _onnxOutputName = MatchTensorName(_session.OutputMetadata.Keys, _outputSpec.Column);
    }

    private static string MatchTensorName(IEnumerable<string> candidates, string preferred)
    {
        var list = candidates.ToList();
        if (list.Count == 1) return list[0];
        var match = list.FirstOrDefault(n => n == preferred || n.StartsWith(preferred + "::") || n.StartsWith(preferred + "_"));
        if (match is null)
        {
            throw new InvalidOperationException(
                $"Could not match an ONNX tensor to feature '{preferred}' among [{string.Join(", ", list)}]. " +
                "Inspect preprocessing.json and adapt this client if the model uses an unrecognized naming scheme.");
        }
        return match;
    }

    private (int Width, int Height) GetImageDims()
    {
        int width = 224, height = 224;
        if (_inputSpec.LudwigPreprocessing is { } pp)
        {
            if (pp.TryGetValue("width", out var w) && w.ValueKind == JsonValueKind.Number) width = w.GetInt32();
            if (pp.TryGetValue("height", out var h) && h.ValueKind == JsonValueKind.Number) height = h.GetInt32();
        }
        return (width, height);
    }

    private DenseTensor<float> PreprocessImage(string path)
    {
        var (width, height) = GetImageDims();
        using var image = Image.Load<Rgb24>(path);
        image.Mutate(x => x.Resize(width, height));

        var norm = _inputSpec.ImageNormalization;
        var data = new float[3 * height * width];
        image.ProcessPixelRows(accessor =>
        {
            for (var y = 0; y < height; y++)
            {
                var row = accessor.GetRowSpan(y);
                for (var x = 0; x < width; x++)
                {
                    var px = row[x];
                    float r = px.R / 255f, g = px.G / 255f, b = px.B / 255f;
                    if (norm is not null)
                    {
                        r = (r - norm.Mean[0]) / norm.Std[0];
                        g = (g - norm.Mean[1]) / norm.Std[1];
                        b = (b - norm.Mean[2]) / norm.Std[2];
                    }
                    data[(0 * height * width) + (y * width) + x] = r;
                    data[(1 * height * width) + (y * width) + x] = g;
                    data[(2 * height * width) + (y * width) + x] = b;
                }
            }
        });
        return new DenseTensor<float>(data, [1, 3, height, width]);
    }

    private static float[] Softmax(float[] logits)
    {
        var max = logits.Max();
        var exp = logits.Select(v => MathF.Exp(v - max)).ToArray();
        var sum = exp.Sum();
        return exp.Select(v => v / sum).ToArray();
    }

    /// <summary>Runs inference and returns class -&gt; probability (classification, sorted descending) or {"value": x} (regression).</summary>
    public Dictionary<string, double> Predict(string imagePath, int topK = 5)
    {
        if (_inputSpec.Type != "image")
        {
            throw new NotSupportedException(
                $"No preprocessing implemented for input type '{_inputSpec.Type}'. " +
                "See preprocessing.json's inputs[0].ludwigPreprocessing and adapt this client.");
        }

        var tensor = PreprocessImage(imagePath);
        var inputs = new List<NamedOnnxValue> { NamedOnnxValue.CreateFromTensor(_onnxInputName, tensor) };
        using var results = _session.Run(inputs);
        var logits = results.First(r => r.Name == _onnxOutputName).AsEnumerable<float>().ToArray();

        var classes = _outputSpec.Classes;
        if (classes is { Count: > 0 } && classes.Count == logits.Length)
        {
            var probs = Softmax(logits);
            return classes
                .Select((name, i) => (name, prob: probs[i]))
                .OrderByDescending(p => p.prob)
                .Take(topK)
                .ToDictionary(p => p.name, p => (double)MathF.Round(p.prob, 4));
        }

        return new Dictionary<string, double> { ["value"] = logits[0] };
    }

    public void Dispose() => _session.Dispose();
}

File: server/lib/export/templates/python/theseus_client.py
"""Theseus-generated inference client.

Wraps the exported ONNX model with the preprocessing/postprocessing Ludwig
applied at training time, read from preprocessing.json — Ludwig's exported
ONNX graph is the bare model only (no resize/normalize, no tokenization, no
class-index decoding), so this client reconstructs that step from the
manifest rather than assuming a fixed tensor layout: input/output tensor
names are matched against the graph's actual names at runtime, not
hardcoded, since ONNX export doesn't guarantee a stable naming convention
across Ludwig versions.

Implements preprocessing for:
  - a single `image` input with a `category`/`number` output (image
    classification/regression) — pass a file path to `predict()`.
  - one or more `number` inputs (tabular classification/regression) — pass
    a `{column: value}` dict to `predict()`, one entry per input feature.

Every other input type (text, audio) raises NotImplementedError with a
pointer to preprocessing.json so you can adapt preprocessing yourself —
text tokenization in particular depends on the chosen encoder's own
vocabulary (Ludwig's built-in tokenizer, or a HuggingFace subword
tokenizer for bert/distilbert/roberta), which can't be generically
reconstructed from preprocessing.json alone. See README.md.
"""

from __future__ import annotations

import json
import math
from pathlib import Path

import numpy as np
import onnxruntime as ort

_HERE = Path(__file__).parent


def _softmax(x: np.ndarray) -> np.ndarray:
    e = np.exp(x - np.max(x))
    return e / e.sum()


class TheseusClient:
    def __init__(
        self,
        model_path: str | Path = _HERE / "model.onnx",
        preprocessing_path: str | Path = _HERE / "preprocessing.json",
    ):
        self.session = ort.InferenceSession(str(model_path))
        with open(preprocessing_path, encoding="utf-8") as f:
            self.manifest = json.load(f)
        self.input_specs: list[dict] = self.manifest["inputs"]
        self.input_spec = self.input_specs[0]
        self.output_spec = self.manifest["outputs"][0]

        onnx_input_names = [i.name for i in self.session.get_inputs()]
        self._onnx_input_names = {
            spec["name"]: self._match_tensor_name(onnx_input_names, spec["column"]) for spec in self.input_specs
        }
        self._onnx_output_name = self._match_tensor_name(
            [o.name for o in self.session.get_outputs()], self.output_spec["column"]
        )
        # Multiple inputs, or a single tabular (number) one, means predict()
        # expects a {column: value} dict rather than a single file path.
        self._is_tabular = len(self.input_specs) > 1 or self.input_spec["type"] == "number"

    @staticmethod
    def _match_tensor_name(candidates: list[str], preferred: str) -> str:
        if len(candidates) == 1:
            return candidates[0]
        for name in candidates:
            if name == preferred or name.startswith((f"{preferred}::", f"{preferred}_")):
                return name
        raise ValueError(
            f"Could not match an ONNX tensor to feature '{preferred}' among {candidates}. "
            "Inspect preprocessing.json and pass the tensor name explicitly if this model "
            "uses a naming scheme this client doesn't recognize."
        )

    def _preprocess_image(self, path: str | Path) -> np.ndarray:
        from PIL import Image

        pp = self.input_spec.get("ludwigPreprocessing") or {}
        height = pp.get("height", 224)
        width = pp.get("width", 224)

        img = Image.open(path).convert("RGB").resize((width, height))
        arr = np.asarray(img, dtype=np.float32) / 255.0  # HWC, [0, 1]

        norm = self.input_spec.get("imageNormalization")
        if norm:
            mean = np.array(norm["mean"], dtype=np.float32)
            std = np.array(norm["std"], dtype=np.float32)
            arr = (arr - mean) / std
        # else: no known normalization preset — raw [0,1] pixels are passed
        # through. Check preprocessing.json's `ludwigPreprocessing` block if
        # predictions look off.

        arr = arr.transpose(2, 0, 1)  # HWC -> CHW
        return arr[np.newaxis, ...].astype(np.float32)

    @staticmethod
    def _preprocess_number(value: float, spec: dict) -> np.ndarray:
        """Replicates the exact normalization Ludwig fit at train time
        (ludwig.features.number_feature.numeric_transformation_registry) —
        the ONNX graph itself has none: NumberInputFeature.forward() goes
        straight from the raw tensor to its encoder.
        """
        x = float(value)
        norm = spec.get("numberNormalization")
        if norm:
            norm_type = norm.get("type")
            if norm_type == "zscore" and norm.get("std"):
                x = (x - norm["mean"]) / norm["std"]
            elif norm_type == "minmax" and norm.get("min") is not None and norm.get("max") is not None:
                span = norm["max"] - norm["min"]
                x = (x - norm["min"]) / span if span else 0.0
            elif norm_type == "log1p":
                x = math.log1p(x)
            elif norm_type == "iq" and norm.get("q2") is not None:
                span = (norm.get("q3") or 0) - (norm.get("q1") or 0)
                x = (x - norm["q2"]) / span if span else x - norm["q2"]
            # Unrecognized normalization type: passed through unchanged —
            # check preprocessing.json's `numberNormalization` if predictions look off.
        # 1-D, shape [1] — matches Ludwig's own create_sample_input() for a
        # single-example number feature (torch.rand([batch_size])).
        return np.array([x], dtype=np.float32)

    def _preprocess_record(self, record: dict[str, float]) -> dict[str, np.ndarray]:
        feed: dict[str, np.ndarray] = {}
        for spec in self.input_specs:
            if spec["type"] != "number":
                raise NotImplementedError(
                    f"No preprocessing implemented for input type '{spec['type']}' in a multi-input model. "
                    "See preprocessing.json and adapt this method."
                )
            if spec["column"] not in record:
                raise KeyError(f"Missing required field '{spec['column']}'")
            feed[self._onnx_input_names[spec["name"]]] = self._preprocess_number(record[spec["column"]], spec)
        return feed

    def predict(self, value: str | Path | dict[str, float], top_k: int = 5) -> dict[str, float]:
        """Run inference and return `{class_name: probability}` sorted
        descending (classification), or `{"value": float}` (regression).

        `value` is a file path for a single `image` input, or a
        `{column: value}` dict — one entry per input feature — for a
        tabular (`number`-only) model.
        """
        if self._is_tabular:
            if not isinstance(value, dict):
                raise TypeError("This model has tabular inputs — pass a dict of {column: value}, not a file path.")
            feed = self._preprocess_record(value)
        elif self.input_spec["type"] == "image":
            feed = {self._onnx_input_names[self.input_spec["name"]]: self._preprocess_image(value)}
        else:
            raise NotImplementedError(
                f"No preprocessing implemented for input type '{self.input_spec['type']}'. "
                "See preprocessing.json's `inputs[0].ludwigPreprocessing` for the "
                "raw parameters Ludwig used and adapt this method."
            )

        raw = self.session.run([self._onnx_output_name], feed)[0]
        logits = np.asarray(raw).reshape(-1)

        classes = self.output_spec.get("classes")
        if classes and len(logits) == len(classes):
            probs = _softmax(logits)
            ranked = sorted(zip(classes, probs), key=lambda p: p[1], reverse=True)
            return {name: round(float(p), 4) for name, p in ranked[:top_k]}

        return {"value": float(logits[0])}
